src/rgw/rgw_sal.h

```diff
@@ -912,23 +900,34 @@ class Zone {
     virtual const string& get_current_period_id() = 0;
 };
 
+class SystemNamespace {
+public:
+  virtual ~SystemNamespace() = default;
+
+  virtual int get(const DoutPrefixProvider *dpp, optional_yield y, const std::string& key, bufferlist& bl) = 0;
+  virtual int put(const DoutPrefixProvider *dpp, optional_yield y, const std::string& key, bufferlist& bl, bool exclusive = false) = 0;
+  virtual int del(const DoutPrefixProvider *dpp, optional_yield y, const std::string& key) = 0;
+  virtual int list(const string& prefix_filter, int max, RGWListRawObjsCtx& ctx,
+		   std::list<string>& oids, bool *is_truncated) = 0;
+};
```

Comment:
this still doesn't feel like the right level of abstraction. my issue with system object isn't that it implies a pool, but that it's based on binary blobs. that puts the encoding above the zipper layer, and prevents the backends from having their own representation or private fields

taking `RGWRole` as an example, we actually store 3 separate system objects - one whose name is `roles.<role_id>` that actually stores the encoded `RGWRole`, and two others that act like symlinks to look up a role by its name and path. a relational database backend would want to put all of that in a single table, with keys for the id, name, and path

this is similar to how rgw users work (they have separate system objects for lookup by access key, email, etc), and the single `rgw::sal::User` interface lets the rados backend manage those objects internally. we should do the same with `rgw::sal::Role`

not all system objects are composites like roles and users, but i still think the database backend would want a separate table for each instead of one table that just has a name and blob

Reply:
I didn't notice this, apologies;  I agree with the above assessment, and have taken this position since last year.  I think the giveaway is these operations taking a ceph buffer::list as an argument.

Reply:
I've been somewhat hesitant to pull other types into Zipper.  I can add Role, I guess, as a full-class Zipper type; but what about the other things using sysobj?  Do have have to make them all full Zipper types?  At what point do we have any classes that are not controlled by zipper?

Reply:
Well, that's a good question.  For the sake of discussion, could we make a list of the types (not yet in zipper) that would need to be there?